### DISCUSSION TITLE
Pruning during replication?

### DISCUSSION THREAD

**Discussion by Jauchi** [zrepl/zrepl#574] [Q&A]
Hello!
I use zrepl to send stuff between my computer (has a raidz1 array) and a NAS that is located offsite.

This works pretty well, although replication is rather slow (2MB/s).

However, with some big changes, GBs start to add up, so it may take a couple of days (or weeks) to catch up with bigger changes.
Since the pruning phase comes strictly **after** replication is completed, this means that a lot of stuff is piling up - even changes that have already been replicated to the remote and would be pruned, I still need to wait (potentially weeks) for pruning to actually take place.
I am aware that I can just create a snap job to remove obsolete snapshots locally, that may however interfere with currently running replications and seems kind of hacky to me.

Would it be possible to prune the receiver and sender either periodically or between steps, so that this kind of buildup could be mitigated somewhat?

**Comment by InsanePrawn** [comment]
In principle, you could offload pruning to a `snap` job (without snapshotting) completely, that'll run async.
deleting snapshots that zrepl had planned to replicate will cause the replication task to fail the run, so it'll need to be rerun.
So try not to configure too small a snapjob interval, and smaller intervals for the replication I suppose. If needed, decouple snapshotting from replication with yet another (or the same, if you're fine with snapping and pruning running at the same intervals) snap job.
You can refer to this for inspiration: https://zrepl.github.io/quickstart/backup_to_external_disk.html

**Reply by Jauchi** [reply]
Hello!
Thank you very much for the quick reply.
I just wrote a little shell script that will delete snapshots up to a certain point - in my case the last snapshot present on both sides.
My question was more related to it being technically possible to prune before or during replication (for example during each step), so that my shell script (or the snap workaround) become unnecessary. The docs say that this it's not currently implemented this way and I wanted to know if it's possible to change that:
>The keep rules are evaluated on the active side (push or pull job) of the replication setup, for both active and passive side, after replication completed or was determined to have failed permanently.

I snapshot every 15 minutes (will probably drop that down to hourly), but breaking the replication every hour does not seem like an ideal solution.

**Reply by problame** [reply]
@Jauchi you wouldn't be "breaking" the replication every hour. It would just fail on _that attempt_, but restart planning when you wake it up again via `zrepl signal wakeup JOB`.

**Reply by Jauchi** [reply]
> @Jauchi you wouldn't be "breaking" the replication every hour. It would just fail on _that attempt_, but restart planning when you wake it up again via `zrepl signal wakeup JOB`.

Sorry, poor choice of words.
My main issue with that snap job approach is that there is not `not_replicated` policy possible, which might affect resolution on the receiving side permanently. It also feels kind of wrong to just delete stuff from other jobs, to be honest.

Also, (please correct me if I'm wrong) waking up is a manual job, so assuming that I don't always check if zrepl is currently having an issue, data is not replicated for (up to) another hour until the next attempt, which will pile additional data onto the amount of data that already needs to be transmitted (because it just sits there until the job auto-retries).
I know that you could just add a cronjob every minute, but that also sounds like it's against design principles.

Is it really not possible to add an option to prune once every _interval_ seconds, or prune after n successful steps?

**Reply by problame** [reply]
@Jauchi sorry for the late reply. 

> Is it really not possible to add an option to prune once every interval seconds, or prune after n successful steps?

You mean, something like [this](https://github.com/zrepl/zrepl/issues/397#issuecomment-766143183), but for pruning?

**Reply by Jauchi** [reply]
Well, that seems like quite a lot of work, since you would then have to presumably have to stop and re-calculate the replication whenever the pruning takes place, correct?
I would be very happy if you could just stop the job (manually), and signal pruning like you could signal starting.

My current workflow is like this:
* I get an alert that my dataset is using lots of storage space
* I determine that the cause is lots of hourlies - which built up since the last successful replication (for example 4 weeks ago)
* Current replication is still ongoing
* These hourlies could be deleted without affecting any pruning policies (They will be deleted once replication is over)
* I then need to calculate which ones would be deleted by zrepl, since it won't do it itself
* Stop the job, delete them.
* Start again, no problems.

And the new workflow could be:
* Get an alert that my dataset is using lots of storage space
* Determine old (useless) snapshot buildup to be the case, and they need to be pruned
* Send a stop signal to the offending zrepl job
* Send a prune signal to the zrepl job
* The zrepl job now prunes, without having successfully completed replication, the old snapshots (ones that were already replicated or are no longer needed)
* I start the job again, with all the snapshots removed and free storage space

Alternatively, a cli argument to print snapshots that would be pruned would be handy. Maybe that would be the easiest to implement?

I actually just need any way to get snapshots that I can remove source and destination side. That would also cut down on replication time, since old useless data does not need to be transmitted anymore.

**Reply by problame** [reply]
> I then need to calculate which ones would be deleted by zrepl, since it won't do it itself

If you don't use the [`not_replicated` rule](https://zrepl.github.io/configuration/prune.html#policy-not-replicated) this shouldn't happen. Do you?

**Reply by Jauchi** [reply]
> If you don't use the [`not_replicated` rule](https://zrepl.github.io/configuration/prune.html#policy-not-replicated) this shouldn't happen. Do you?

Hello, my issue is that I can't run `zrepl --show-snapshots-to-destroy` or something. zrepl currently can't print the snapshot names I should use with `zfs destroy`, so that's what I meant with `zrepl won't do it itself `.
Could have worded this better, sorry!
